### PR TITLE
[QACI-410] - Classloader Leak Tracking

### PR DESCRIPTION
**Pull request by cubastanley**
## Description
Technically a new feature - adds a new sample app that allows for tracking the number of instances of WebappClassLoader before and after a series of deployments

### New tests
Currently removed classloader leak test in samples - removed due to intergration blockers

### Testing Performed
Manual testing, deployed locally and returns an appropriate value for instance count

### Testing Environment
OSX, Zulu JDK 1.8.0_272

## Documentation
README included with usage instructions

## Notes for Reviewers
This is an internal tool as it were, a step in the direction of being able to test for Classloader instance leaks (i.e. instances of WebappClassLoader that aren't garbage collected when they should be) 

Excuse the formatting in the Samples pom - NetBeans decided it would be fun to insert a new line between every line in the file

### CODE REVIEW

File: appserver/tests/payara-samples/classloader-data-api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    <name>Payara Samples - ClassLoader Data API</name>
+    <description>A REST API Application for retrieving an instance count of 
+        WebappClassLoader from an instance of Payara Server</description>
+    
+    <properties>
+        <jersey.version>3.0.0</jersey.version>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+    
+    <dependencyManagement>
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/pom.xml:25]: no need for Jersey, use plain jax-js APIs instead

File: appserver/tests/payara-samples/classloader-data-api/pom.xml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                <version>${jersey.version}</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+    
+    <dependencies>
+        <dependency>
+            <groupId>org.glassfish.jersey.containers</groupId>
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/pom.xml:39]: same here, use plain jax-rs APIs

File: appserver/tests/payara-samples/classloader-data-api/pom.xml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        <!-- Payara Dependencies -->
+        <dependency>
+            <groupId>fish.payara.server.internal.web</groupId>
+            <artifactId>war-util</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+    
+    <build>
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/pom.xml:56]: no need for this section at all, use properties equivalent
This way you don't override the complier plugin version at all

File: appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountService.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountService.java:1]: full Payara copyright

File: appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountService.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package fish.payara.samples.classloaderdata;
+
+/**
+ *
+ * @author cuba
+ */
+public class InstanceCountService {
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountService.java:12]: Not sure this is necessary at all

File: appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountTracker.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.lang.ref.Reference;
+import org.glassfish.web.loader.WebappClassLoader;
+import org.glassfish.web.loader.WebappClassLoaderFinalizer;
+
+/**
+ *
+ * @author Cuba Stanley
+ */
+public class InstanceCountTracker {
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountTracker.java:16]: I would move instance count into own class, not have instance counts in different classes.
Minimize changes to WebappClassLoader

Also, this doesn't touch non-webapp (jar, ear etc) class loaders

File: appserver/tests/payara-samples/classloader-data-api/src/main/webapp/WEB-INF/web.xml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/src/main/webapp/WEB-INF/web.xml:1]: no need for this file, use plain jax-rs APIs

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java [hunk 8]
```diff
@@ -408,9 +409,25 @@
     private static final boolean IS_JDK_VERSION_HIGHER_THAN_8 = JDK.getMajor() > 8;
     private static Boolean isMultiReleaseJar;
     private static final Name MULTI_RELEASE = new Name("Multi-Release");
+    
+    private static int INSTANCE_COUNT;
```
Comment (lprimak) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:413]: possible race condition, should be AtomicInteger

File: appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountTracker.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    
+    private static int previousInstanceCount;
+    
+    public static int getInstanceCount() {
+        System.gc();
+        Reference<? extends WebappClassLoader> referenceFromQueue;
+        while((referenceFromQueue = WebappClassLoader.referenceQueue.poll()) != null) {
+            ((WebappClassLoaderFinalizer)referenceFromQueue).cleanupAction();
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountTracker.java:24]: this cast is a code smell, this needs to be refactored

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java [hunk 10]
```diff
@@ -1,7 +1,7 @@
 /*
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
  *
- * Copyright (c) 1997-2019 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997-2020 Oracle and/or its affiliates. All rights reserved.
```
Comment (lprimak) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:4]: oracle copyrights should not be changed

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoaderFinalizer.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+// Portions Copyright [2020] [Payara Foundation and/or its affiliates]
```
Comment (lprimak) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoaderFinalizer.java:1]: needs full copyright, not just portions

**Review by lprimak — CHANGES_REQUESTED** [review]

File: appserver/tests/payara-samples/pom.xml [hunk 12]
```diff
@@ -37,60 +37,56 @@
  *    and therefore, elected the GPL Version 2 license, then the option applies
  *    only if the new code is made subject to such option by the copyright
  *    holder.
--->
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+--><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (lprimak) [inline appserver/tests/payara-samples/pom.xml:40]: I would just revert and start this file over again

File: appserver/tests/payara-samples/classloader-data-api/pom.xml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    <artifactId>classloader-data-api</artifactId>
+    <version>5.2020.8-SNAPSHOT</version>
+    
+    <packaging>war</packaging>
+    <name>Payara Samples - ClassLoader Data API</name>
+    <description>A REST API Application for retrieving an instance count of 
+        WebappClassLoader from an instance of Payara Server</description>
+    
+    <properties>
+        <jersey.version>3.0.0</jersey.version>
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/pom.xml:21]: remove jersey

File: appserver/tests/payara-samples/classloader-data-api/pom.xml [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    </parent>
+    
+    <groupId>fish.payara.samples</groupId>
+    <artifactId>classloader-data-api</artifactId>
+    <version>5.2020.8-SNAPSHOT</version>
+    
+    <packaging>war</packaging>
+    <name>Payara Samples - ClassLoader Data API</name>
+    <description>A REST API Application for retrieving an instance count of 
+        WebappClassLoader from an instance of Payara Server</description>
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/pom.xml:18]: Need the same functionality for ASURLClassLoader (parent of EARClassLoader EJBClassLoader etc)

File: appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountService.java (hunk 5 above)
Comment (lprimak) [reply appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountService.java:46]: I still see no point in this class, it does nothing.
InstanceCountTracker can be used directly.

File: appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountTracker.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ *
+ * @author Cuba Stanley
+ */
+public class InstanceCountTracker {
+    
+    private static int previousInstanceCount;
+    
+    public static int getInstanceCount() {
```
Comment (lprimak) [inline appserver/tests/payara-samples/classloader-data-api/src/main/java/fish/payara/samples/classloaderdata/InstanceCountTracker.java:53]: needs intent javadoc

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java [hunk 16]
```diff
@@ -409,8 +411,14 @@
     private static Boolean isMultiReleaseJar;
     private static final Name MULTI_RELEASE = new Name("Multi-Release");
 
-    static {
+    public static ReferenceQueue<WebappClassLoader> referenceQueue = new ReferenceQueue<WebappClassLoader>();
```
Comment (lprimak) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:414]: should be final

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java [hunk 17]
```diff
@@ -409,8 +411,14 @@
...
+    public static ArrayList<WebappClassLoaderFinalizer> list = new ArrayList<WebappClassLoaderFinalizer>();
```
Comment (lprimak) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:415]: should be final

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java [hunk 18]
```diff
@@ -409,8 +411,14 @@
...
+    {
```
Comment (lprimak) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:416]: is isn't really standard, you should move it down into the `static` block

**Review by lprimak — CHANGES_REQUESTED** [review]
It was good, now much, much better. Just minor things left

File: appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java [hunk 19]
```diff
@@ -1887,8 +1894,11 @@ protected PermissionCollection getPermissions(CodeSource codeSource) {
 
 
     private void init() {
-
+        
         this.parent = getParent();
+        
+        list.add(new PhantomReference(this, referenceQueue));
+        INSTANCE_COUNT.set(INSTANCE_COUNT.get() + 1);
```
Comment (sgflt) [inline appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java:1901]: This does not seem to be thread safe too.

**Review by lprimak — CHANGES_REQUESTED** [review]
There are code duplication smells still in the code, with duplicated parts in ASURLClassLaoder vs. WebappClassLoader. This can be refactored into another lass

**Review by lprimak — APPROVED** [review]

**Comment by lprimak** [comment]
jenkins test

File: nucleus/common/common-util/src/main/java/org/glassfish/common/util/InstanceCounter.java [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        public Count() {
+            instanceCount = 1;
+        }
+
+        public Count(int instanceCount) {
+            this.instanceCount = instanceCount;
+        }
+    }
+
+    private static final Map<Class<?>, Count> INSTANCE_COUNT = new ConcurrentHashMap<>();
```
Comment (sgflt) [inline nucleus/common/common-util/src/main/java/org/glassfish/common/util/InstanceCounter.java:77]: I would use ConcurrentMap interface to point that there is expected concurrency.

```suggestion
    private static final ConcurrentMap<Class<?>, Count> INSTANCE_COUNT = new ConcurrentHashMap<>();
```

**Comment by lprimak** [comment]
Jenkins test

**Comment by lprimak** [comment]
jenkins test

**Pull request merged**